File: maxent/scripts/charutil.py
# -*- coding: utf-8 -*-
"""
字符 工具类，用来判断和处理字符
"""

#判断是否是数字
digits_str=u'1234567890〇一二三四五六七八九十'
digits= set(list(digits_str))
def isdigit(char):
	return char in digits

#全角字符转换成半角字符
def strq2b(ustring):
	rstring = ""
	for uchar in ustring:
		inside_code=ord(uchar)
		if inside_code == 12288:#全角空格直接转换
			inside_code = 32
		elif inside_code>=65281 and inside_code<=65374:
			inside_code-=65248
		rstring += unichr(inside_code)
	return rstring

#判断是否是标点符号
#。？！，、；：“”‘’（）{}【】—…《》
asc_punc_str=u"!\"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~"
chi_punc_str=u"。？！，、；：“”‘’（）{}【】—…《》"
punc=set(list(asc_punc_str+chi_punc_str))
def ispunc(char):
	return char in punc

#是否是日期单位: 年/月/日
dateunits=set(list(u'年月日'))
def isdateunit(char):
	return char in dateunits

#是否是英文字符:大写、小写 共52个
def isletter(char):
	if (char>='a' and char<='z') or (char>='A' and char<='Z'):
		return True
	return False
	
#是否是汉字
#\u4e00-\u9fa5
#\uF900-\uFA2D
def ishanzi(char):
	if (char >= u'\u4e00' and char <= u'\u9fa5') or (char >= u'\uF900' and char <= u'\uFA2D'):
		return True
	return False 

File: maxent/src/opennlp/maxent/CharUtil.java
package opennlp.maxent;

import java.util.HashSet;

public class CharUtil {
	 static HashSet<Character> digits=new HashSet<Character>();
	 static HashSet<Character> punc = new HashSet<Character>();
	 static{
		 String digits_str="1234567890〇一二三四五六七八九十";
		 for(int i=0,len=digits_str.length();i<len;i++){
			 digits.add(digits_str.charAt(i));
		 }
		 String punc_str="!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~。？！，、；：“”‘’（）{}【】—…《》";
		 for(int i=0,len=punc_str.length();i<len;i++){
			 punc.add(punc_str.charAt(i));
		 }
	 }
	 
	 
	// 1234567890〇一二三四五六七八九十
	public static boolean isdigit(char ch){
		return digits.contains(ch);
	}
	
	//全角转半角
	public static String strq2b(String text){
		return String.valueOf(strq2b(text.toCharArray()));
	}
	
	public static char[] strq2b(char[] text){
		char[] newText = new char[text.length];
		for(int i=0;i<text.length;i++){
			char ch=text[i];
			if(ch==12288)
				newText[i]=' ';
			else if(ch>=65281 && ch<=65374){
				ch-=65248;
				newText[i]=ch;
			}else{
				newText[i]=ch;
			}
		}
		return newText;
	}
	
//	#判断是否是标点符号
//	#。？！，、；：“”‘’（）{}【】—…《》
//	asc_punc_str=u"!\"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~"
//	chi_punc_str=u"。？！，、；：“”‘’（）{}【】—…《》"
	public static boolean ispunc(char ch){
		return punc.contains(ch);
	}
	//是否是日期单位: 年/月/日
	public static boolean isdateunit(char ch){
		return ch=='年'||ch=='月'||ch=='日';
	}
	//是否是英文字符:大写、小写 共52个
	public static boolean isletter(char ch){
		if( (ch>='a' && ch<='z') || (ch>='A' && ch<='Z'))
			return true;
		return false;
	}
	//是否是汉字
	//\u4e00-\u9fa5
	//\uF900-\uFA2D
	public static boolean ishanzi(char ch){
		if ((ch >= '\u4e00' && ch <= '\u9fa5') || (ch >= '\uF900' && ch <= '\uFA2D'))
			return true;
		return false; 
	}
	
	
}
